### ISSUE TITLE
mtr shows false loss when path changes in the middle of execution

### ISSUE THREAD

**Issue by atibdialpad** [traviscross/mtr#361]
For example, the destination was first reachable in 5 hops. Somewhere during RUN-2, path changed and the dest started being reachable in 8 hops. I guess, mtr continued probing with max_ttl of 5 (since it was previously reachable with ttl=5) and declared losses.

**RUN -1 : /usr/local/sbin/mtr -4 --report-wide --report-cycles 50 170.10.213.41**
Start: 2020-07-03T12:16:25+0000
HOST: tco-jhb-engine4                        Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 170.10.214.1                            0.0%    50    1.8   2.9   0.7  13.0   3.1
  2.|-- xe-0-0-1-0.tco-jhb-ebr1.ubermanage.net  0.0%    50    0.3   2.1   0.2  44.5   7.8
  3.|-- 185.91.10.131                           8.0%    50  163.2 165.0 163.1 215.0   8.2
  4.|-- xe-0-2-0.ctl-sao-asw1.ubermanage.net    8.0%    50  163.9 164.6 163.7 170.3   1.0
  5.|-- 170.10.213.41                           8.0%    50  163.3 163.3 163.2 163.4   0.1

**RUN -2 : /usr/local/sbin/mtr -4 --report-wide --report-cycles 50 170.10.213.41**
Start: 2020-07-03T12:17:30+0000
HOST: tco-jhb-engine4                        Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 170.10.214.1                            0.0%    50    1.4   1.5   0.7   2.9   0.6
  2.|-- xe-0-0-1-0.tco-jhb-ebr1.ubermanage.net  0.0%    50    0.3   2.4   0.2  46.9   8.5
  3.|-- 185.91.10.131                          20.0%    50  182.9 184.9 163.1 230.2  11.2
  4.|-- xe-0-2-0.ctl-sao-asw1.ubermanage.net   20.0%    50  256.8 255.1 163.8 273.2  15.1
  5.|-- 170.10.213.41                          20.0%    50  257.0 257.9 163.3 318.9  19.1
 
**RUN -3 : /usr/local/sbin/mtr -4 --report-wide --report-cycles 50 170.10.213.41**
Start: 2020-07-03T12:18:38+0000
HOST: tco-jhb-engine4                             Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 170.10.214.1                                 0.0%    50    0.8   1.7   0.7  19.7   2.6
  2.|-- xe-0-0-1-0.tco-jhb-ebr1.ubermanage.net       0.0%    50   41.6   2.3   0.2  41.6   7.4
  3.|-- ge-0-0-3.407.eqx-ams-ebr1.ubermanage.net     0.0%    50  182.6 185.4 182.5 240.0  11.5
  4.|-- xe-0-1-1-0.1509.drt-nyc-ebr2.ubermanage.net  0.0%    50  256.9 259.9 256.7 303.0  10.0
  5.|-- et-0-1-0.drt-nyc-ebr1.ubermanage.net         0.0%    50  256.9 259.9 256.7 323.4  11.2
  6.|-- ge-0-0-5.ctl-sao-ebr1.ubermanage.net         0.0%    50  371.0 373.2 370.8 415.2   8.8
  7.|-- xe-0-2-0.ctl-sao-asw1.ubermanage.net         0.0%    50  371.5 372.3 371.4 373.2   0.6
  8.|-- 170.10.213.41                                0.0%    50  370.8 370.9 370.8 371.2   0.1

Expecatation : When path changes in the middle, can mtr start probing with higher TTL and output something like :  
"For packet 1-20, <path_info>
Then path changed
New Path info for packets <21-50> "

Just want to confirm if this is a known issue / issue at all / can be solved by some existing flags etc.
Happy to learn the code and make the changes if you think this makes sense

Just to give more context, some machines (esp ones inside Gloud Network like GCP, AWS) have a very dynamic network path and these false losses sends out many spurious alerts. 
@rewolff

**Comment by rewolff** [comment]
The problem I see is that it is not easy to see a "permanent change in route": When you get the first packet back from someone you don't expect, then you don't know if this is just "hop 5 can be served both by XXX and YYY" or a permanent change. 
Also,  on the display side: it would be nice to be able to show (as in this case!) route abcdef and route abc-klmnop and that abcdef was last active X seconds ago. 
Implementation: not sure how to store such information correctly that if you leave something like that running for a day that you can show 40% of the time we saw abcdef and 60% of the time it was abcklmnop. That would be ideal. 
My plan-of-attack would be to start with: how you would want  to present the information that is possible in practice and then work on the datatypes that would be able to store the required info and THEN start working on the implementation. 
You already started with the "i'd like it to show", but that needs to be expanded a bit. Also keep in mind that especially the "report" output mode might be parsed by programs.... In the simple cases it would be nice to remain compatible.

**Comment by atibdialpad** [comment]
Thanks @rewolff for looking at it and suggestions. Going thru the code to come up with a proposal. 
Another point I wanted to make was I always see these false errors when path elongates mid way (for example in the above runs, path increased from 5 to 8). I never see MTR reporting losses when the path recovers (8->5)

**Comment by rewolff** [comment]
About the path elongation: 
When I initially startted with Mtr, it would always probe upto 20 hops away. Then, that max 20 probe depth is reduced to where the target is when we get the reply that indicates that we reached the target. Nowadays, mtr probes to "5 further than the currently furthest reply that we got", of course unless that's the target. 
I'm not sure if that max probe depth can increase when the path increases say from 5 to 8 hops. 

again, first, I suggest you think about how to display a situation where a 5 hop route works for an hour (3600 round trip packets to each of the 5 hops) into an 8 hop route, say now 1200 seconds ago. 

Next, the change is the other way around. 

Next, it changes one way, but then swaps back again after 500 cycles. 

What would be a display of that information that we humans can read. 

Maybe we should consider the "current path" the "active path" and display that. and allow people to "ask for" the alternative paths that existed in the past. So once there is a switchover from abcde to abklmnop we just display the abklmnop one and note that "press X" to see another path that existed in the past. 

This may become a bit flipflop when there is abcdef and abCdef where most of the path is the same just a single alternate in the middle...  How to handle that. And what about abcdef vs abCDef ?

**Comment by atibdialpad** [comment]
Initial thought I had on how to display was like :

* a new flag where a user could say that they "WANT" path change information
* a limitation on how many such path changes could be displayed
./mtr d -c 10 —show_path_changes
<max path change that can be tracked, for example 5>

**SAMPLE 1 : Path Elongates after 4 packets**
Packet 1-4  : abcd
Packet 5-10 : abclmnod
 
Hop num      : <Hop  loss%. Snt last avg best worst stdev>
Hop 1.|--    :   a     0%   10   -    -    -    -    -
Hop 2.|--    :   b     0%   10   -    -    -    -    -
Hop 3.|--    :   c     0%   10   -    -    -    -    -
Hop 4.|--    :   d     0%   4    -    -    -    -    -    |  l     0%   6    -    -    -    -    -
Hop 5.|--    :                                                     |  n     0%   6    -    -    -    -    -  
Hop 6.|—     :                                                    |  o     0%   6    -    -    -    -    - 
Hop 7.|—     :                                                    |  d     0%   6    -    -    -    -    - 

<Could display some final summary for DESTINATION hop ‘d’ as well for program which will try to find hop ‘d’ and parse the stats>

**SAMPLE 2 : Path Changes after 4 packets but total length remains same**
Packet 1-4  : abcd
Packet 5-10 : abCd


Hop num      : <Hop  loss%. Snt last avg best worst stdev>
Hop 1.|--    :   a     0%   10   -    -    -    -    -
Hop 2.|--    :   b     0%   10   -    -    -    -    -
Hop 3.|--    :   c     0%   4    -    -    -    -    -  | C    0%   6    -    -    -    -    -
Hop 4.|--    :   d     0%   4    -    -    -    -    -

**SAMPLE 3 : Path Changes after 4 packets but total length remains same.**
Packet 1-4 : abcdef
Packet 4-10: abCDef

Hop num      : <Hop  loss%. Snt last avg best worst stdev>
Hop 1.|--    :   a     0%   10   -    -    -    -    -
Hop 2.|--    :   b     0%   10   -    -    -    -    -
Hop 3.|--    :   c     0%   4    -    -    -    -    -  | C    0%   6    -    -    -    -    -
Hop 4.|--    :   d     0%   4    -    -    -    -    -  | D    0%   6    -    -    -    -    -
Hop 5.|--    :                                                   | e    0%   4    -    -    -    -    -
Hop 6.|--    :                                                   | f    0%   4    -    -    -    -    - 


**SAMPLE 4 : Path Changes after 4 packets but comes back to original config for the last 4 packets**
Packet 1-4  : abcd
Packet 5-6 : abclmnod
Packet 8-10  : abcd

Hop num      : <Hop  loss%. Snt last avg best worst stdev>
Hop 1.|--    :   a     0%   10   -    -    -    -    -
Hop 2.|--    :   b     0%   10   -    -    -    -    -
Hop 3.|--    :   c     0%   10   -    -    -    -    -
Hop 4.|--    :   d     0%   4    -    -    -    -    -   | l    0%   2    -    -    -    -    - | d     0%   2    -    -    -    -    - 
Hop 5.|--    :                                                    | m    0%   2    -    -    -    -    -  
Hop 6.|--    :                                                    | n    0%   2    -    -    -    -    - 
Hop 7.|--    :                                                     | o    0%   2    -    -    -    -    - 
Hop 8.|--    :                                                    | d    0%   2    -    -    -    -    -

**Comment by atibdialpad** [comment]
When I first thought about it, I was thinking of a display like this. In addition, I would also have a filed wherein a user can OPT to TRACK PATH CHANGES and there would also be an upper limit on how many such changes the tool can track.

<img width="1055" alt="Screenshot 2020-08-05 at 4 31 45 PM" src="https://user-images.githubusercontent.com/48474281/89406074-370e9900-d73a-11ea-8e3a-9838ded76fb7.png">

**Comment by atibdialpad** [comment]
When I first thought about it, I was thinking of a display like this. In addition, I would also have a field wherein a user can OPT to TRACK PATH CHANGES and there would also be an upper limit on how many such changes the tool can track.

<img width="1038" alt="Screenshot 2020-08-05 at 4 59 16 PM" src="https://user-images.githubusercontent.com/48474281/89407797-0bd97900-d73d-11ea-944f-890af8355828.png">

**Comment by atibdialpad** [comment]
@rewolff any thoughts on the proposed representations ?

**Comment by atibdialpad** [comment]
I did a deep dive on the code and found answers to some of the questions :
1) Why path elongation is reported as a loss ?
Ans : When mtr is able to reach the destination (lets say at HOP-10) it restricts max_ttl for all future probes to 10. https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L585, so if suddenly the path elongates and the destination is now reachable at HOP-15, mtr probes's TTL will expire well before it can reach the destination. 
On the report/display side, HOP-10 will only show the stats of all probes that were able to meet the destination at HOP-10 (before path elongated) and hence report shows loss.

2) Why path shrinking is not reported as a loss ?
Ans : This has to do with how MTR code is designed. From going thru the code I realized that a line like this :  
3.|-- A.B.C.D 0.0% 50 182.9 184.9 163.1 230.2 11.2
means that "all packets sent with TTL-3 were received back and A.B.C.D is the latest/recent IP to respond from hop-3" and not "A.B.C.D is the only IP at HOP-3 and this particular IP has responded for all the 50 packets sent for HOP-3"

What this means is, says the path was initially src->a->b->c->dst and we were running mtr with count  =50
suppose for the first 30 packets, the path remained constant and stats looked like :  
src.|--0% 30
a.|--   0% 30
b.|-   -0% 30
c.|--   0% 30
dst.|--0% 30
Now, say the path shortens to src->a->b->dst,Now, the 'dst' is the latest IP to respond for TTL-4 packets (for the remaining 20 packets) and so the final stats will look like : 
src.|--0% 50
a.|--   0% 50
b.|-   -0% 50
dst.|--0% 50

If you were to look at the internal data structure, it looks like
HOP-1.  nh = src, multipath_nhs = [src]
HOP-2. nh = a, multipath_nhs = [a] 
HOP-3. nh = b, multipath_nhs = [b]
HOP-4 nh = dst, multipath_nhs = [c, dst].  <--latest to respond is 'dst' and previously it was 'c'
HOP-5 nh = dst, multipath_nhs = [dst]

So since HOP-4 responded back for all 50 packets, it shows as 0% loss but it hides the fact that the path changed mid way and reports the latest path.

**PROPOSAL**:
What I want to do is in https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L210
where we are processing returned packets and populating the internal data structure, we can detect path_changes with something like :  

Using sams ds names as used in the code:  
addr : current respinding src for ttl = index  + 1
nh_addr[index] : last host that responded for ttl = index + 1
remote_address : dst
unspec_addr : 0.0.0.0

// PATH ELONGATION
if (addr != unspec_addr && nh_addr[index] == remote_address && addr != remote_address) {
  // remote was reachable at ttl=index+1 previously but not any more
  ctl.path_change = 1;
}
// PATH SHRINKING
if (addr == remote_address && nh_addr[index] != remote_address && nh_addr[index]!= unspec_addr ) {
// remote is suddenly reachable at ttl=index+1 but was previously not. Some other IP was responding at this TTL.
path_change = 1;
}
In either case, I will set a boolean flag in mtr context 'mtr->path_changed = True' and return to main to restart the probe process arefresh.

Add an option detect_path_change -> User needs to explicitly mention that she/he wants path changes to be tracked
Add a flag : path_changed

Main today looks like :  

int main() {
    <some_stuff>
    display_open(&ctl);
    display_loop(&ctl);
    net_end_transit();
    display_close(&ctl);
    <some_stuff>
}

It can be changed to:  
int main() {
    //same some stuff
    do {
        if (ctl.path_changed) {
            ctl.path_changed = 0; //Reset Flag
            net_reopen();               //Reset internal Data Structure
        }
        display_open(&ctl);
        display_loop(&ctl);          //<--This sets ctl.path_change = 1 if path change is detected
        net_end_transit();
        display_close(&ctl);
    } while (ctl->detect_path_change && ctl->path_changed);
    <same some stuff>
}

This doesn't change any thing if some one doesn't explicitly puts the options to track path changes, making it perfectly compatible with existing programs that depend on output format.

@rewolff please let me know what you think of this.

**Comment by rewolff** [comment]
OK. Good work. About how the path elongation and shortening work: Yes, that was my impression too. 

About your proposal: That would "throw away" all the old info when a path change is detected. So when you start mtr on day one at 8AM and come back the next day at 8AM, you expect to see about 24*3600 = 86400 probes. But now mtr shows just 3600 probes! (The route changed back at 7AM). So... I would say we try to not throw away the old info.

If for example, we add an mtr-time-stamp (round number? probe number?) to each host at positionX is y record about when it was last seen to be active, then we can gather the information into the datastructure, retain enough info to show the "active route" and it is up to the display functions to present the info in a readable way. 

The "path changed" boolean would not be necessary, but your "if" code before that still should detect those. And take appropriate action. For the host at index is not destination as expected, that would be maxhops++, and for the other one a maxhops--;

**Comment by atibdialpad** [comment]
Thanks for taking the time to go through the analysis.

I did initially think about aggregating all data in hosts[] and leaving it to the report() function to decipher the last active and current active paths. However I realised it would need lot of changes in the report() function and would also become complicated when there are multiple path changes. We would also need to handle ECMP cases here so that they don't get into the "path-length" handling bucket.
I would again go through this approach and brainstorm once again though :-) 

However, In the proposal above we am not throwing away the old result. Once I detect a path-change, a) I print/report the old stats b) Log a line that path has changed c) reset the internal DS(net_reset) d) and start again.
This automatically takes cares of print OLD and RECENT path information w/o having to make any significant change in the code.
Something like:
a 0% 30..
b 0% 30..
c 0% 30..
d 0% 30..
MTR DETECTED PATH CHANGE..
a 0% 50..
b 0% 50..
c 0% 50..
x 0% 50..
d 0% 50..
I can restart the mtr on path-change with the same count or the remainder count (example, 20 in this example)

**Comment by rewolff** [comment]
On Thu, Aug 20, 2020 at 03:05:40AM -0700, atib wrote:
> This automatically takes cares of print OLD and RECENT path
> information w/o having to make any significant change in the code.

I think you're thinking of a long-running mtr -r report. That has the
option to simply report what it's got and then start over with the new
route.

At least two output options have a "live display" of what's going on.
Although I do hear about people leaving the report version running for
long times, I can also imagine that when users complain about
connectivity issues a sysadmin (or savvy user) would start a few mtr
runs in interactive mode, and then when time comes to go to bed or
leave the office, then they decide to simply let it run and see what
it shows in the morning. So for those cases, a change in the output
routines would be necessary.

As a way of partitioning the project, I would recommend doing the
minimal changes to say show only the current path in the interactive
display modes, and then later expand the interactive displays ot also
show the older paths.

**Comment by atibdialpad** [comment]
That is correct. To start, I want to add path_change support to the report_mode w/o changing anything in interactive/live display modes and this can be done with minimal changes to existing code by using the proposed technique of i) report pre path_change stats ii) Log a msg iii) report post path_change stats
The interactive and other displays like json/xml/raw etc will continue to function as is and can be later expanded to detect path_changes.

**Comment by rewolff** [comment]
Oh... I was under the impression that "net reopen" would cause a reset of existing data.

**Comment by atibdialpad** [comment]
Yes it does.
I am invoking it after display_close() which would have already printed the pre path_change stats. net_reopen() will be called in the do_while loop before the post path_change mtr starts.

**Comment by atibdialpad** [comment]
hmm, I get your point on why printing multiple reports on path-changes is not a very good way to report the result.
You made a very good point of not trying to decipher paths but concentrate on reporting all the hosts that have replied from a HOP at distance Y and also showing what the latest host is at that distance.

https://github.com/traviscross/mtr/pull/362 this change that I made couple of weeks back already takes care of printing all hosts that have replied for a particular hop count, so we are good on that front. All we need is a way to dynamically change MaxTTL of future probes when path changes.

This can be done simply if we 1) Always store the latest responder in nh->addr and previous responders to nh->addrs[] in https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L210
This way we not allow a hard reset of TTL here : https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L580

The output will look like this. 
1) Path Elongation : Say, path was src->a->b->c->dst and changed to a->b->c->d->dst
HOP 1 - a
HOP 2 - b
HOP 3 - c
HOP 4 - d
-------- [d,
---------dst]  <--- Says that 'd' is the latest host at HOP-4 and bucket of all hosts seen at HOP-4 are [d, dst,... ]
HOP 5 - dst

1) Path Shortening : Say, path was src->a->b->c->d->dst and changed to a->b->c->dst
HOP 1 - a
HOP 2 - b
HOP 3 - c
HOP 4 - dst
-------- [dst,
--------  d] <---Says tht 'dst' is the latest host at HOP-4 and bucket of all hosts seen at HOP-4 are [dst, d,...]
HOP 5 - dst

In general, the entry next to 'HOP X' will be the latest responder from a distance X followed by a list of all hosts seen at that distance.

**Comment by rewolff** [comment]
In the old days, when I first started working on mtr, it would always send out probes upto 30 hops deep. That's wasteful of network bandwidth, so I suggest that when we find the destination at hopX we set the maxTTL to that value. 
On the other hand, when we find a "not-the-destination" at the maxTTL hop, I suggest we increase maxTTL, maybe even by two. So if a route lengthens by 5 hops, we are fully probing the path after a reasonable amount of time.

**Comment by atibdialpad** [comment]
Right, that is exactly what will happen. The change to "always store the latest hop at distance X in host[X]" will take care of this. 

Imagine, the path was SRC->A->B->C->DST
The data would look like this:
host.     addr.   addrs
Host 0 SRC.    []
Host 1.   A.       []
Host 2.   B.      []
Host 3.   C.      []
Host 4.  DST.   []

and we continue to probe with max TTL of 5 (since dst was last seen at index 4) because of this check https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L580

Now say path elongated, so we will get a different host at hop-5. we will immediately write host[4] = NEW_HOST (since we always write the latest host). 
host.     addr.   addrs
Host 0 SRC.    []
Host 1.   A.       []
Host 2.   B.      []
Host 3.   C.      []
Host 4.  NEW_HOST.   [DST]
This will prevent https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L580 from limiting the ttl in the very next round and we continue to probe till we reach DST (just like how the first round always work)

Now say, the DST is reachable at Host[10], the same code will set maxTTL to 10 so that subsequent probes go only till 10.

In this way, we don't even have to detect path-changes, the code takes care of ++/-- TTL to where the DST was last seen.

**Comment by atibdialpad** [comment]
I tested out this piece of code and results were quite good 
https://github.com/traviscross/mtr/compare/master...atibdialpad:custom-mtr
**1) Path Elongation : Normal path is a 5 HOP path:** 
```
root@eqx-sjc-engine1-staging:~/atib_custom_mtr/mtr# ./mtr 103.35.125.43 -n --report-wide -c 50
Start: 2020-08-25T10:23:16+0000
HOST: eqx-sjc-engine1-staging Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 103.35.127.1             0.0%    50    2.2   4.5   0.8  57.2   8.7
  2.|-- 185.91.11.98             0.0%    50    0.3   0.9   0.3  15.7   2.5
  3.|-- 185.91.10.122            0.0%    50   48.4  48.9  48.4  60.7   2.0
  4.|-- 185.91.11.161            0.0%    50   51.8  49.9  48.9  56.3   1.3
  5.|-- 103.35.125.43            0.0%    50   48.4  48.4  48.3  48.4   0.0
```
When Path Elongates mid-way of mtr and become a 6 HOP path
```
root@eqx-sjc-engine1-staging:~/atib_custom_mtr/mtr# ./mtr 103.35.125.43 -n --report-wide -c 100
Start: 2020-08-25T10:29:42+0000
HOST: eqx-sjc-engine1-staging Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 103.35.127.1             0.0%   100    4.4   2.0   0.8  15.7   2.3
  2.|-- 185.91.11.98             0.0%   100    0.3   1.0   0.2  18.1   2.9
  3.|-- 185.91.10.118            0.0%   100   36.4  44.3  36.3  75.7   6.7
        185.91.10.122          
  4.|-- 185.91.10.120            1.0%   100   59.2  53.3  48.8  66.6   5.1
        185.91.11.161          
  5.|-- 185.91.11.161            1.0%   100   60.9  52.8  48.4  61.3   5.8
        103.35.125.43          
  6.|-- 103.35.125.43            0.0%    37   59.0  58.8  48.4  59.1   1.8
```
There's enough information here to deduce :
1. Current path is a 6 HOP path since dst (103.35.125.43 ) is the Primary Host at HOP - 6
2. Old path was a 5 HOP path since dst(103.35.125.43 ) is a Secondary Host at HOP - 5
3. Path Changed after 63 (100-37) packets were sent

**Case 2: Path Shortening : Normal Path is a 6 HOP Path**
```
root@eqx-sjc-engine1-staging:~/atib_custom_mtr/mtr# ./mtr 103.35.125.43 -n --report-wide -c 50
Start: 2020-08-25T10:24:53+0000
HOST: eqx-sjc-engine1-staging Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 103.35.127.1             0.0%    50    2.1   1.8   0.8   9.7   1.3
  2.|-- 185.91.11.98             0.0%    50    0.3   0.9   0.2  15.3   2.6
  3.|-- 185.91.10.118            0.0%    50   36.4  36.9  36.3  47.7   2.1
  4.|-- 185.91.10.120            0.0%    50   59.2  59.3  59.1  66.2   1.0
  5.|-- 185.91.11.161            0.0%    50   60.8  60.4  59.6  64.1   0.8
  6.|-- 103.35.125.43            0.0%    50   59.0  59.0  59.0  59.1   0.0
```
Path Shortens to 5 HOPs
```
root@eqx-sjc-engine1-staging:~/atib_custom_mtr/mtr# ./mtr 103.35.125.43 -n --report-wide -c 100
Start: 2020-08-25T10:26:03+0000
HOST: eqx-sjc-engine1-staging Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 103.35.127.1             0.0%   100    0.8   1.5   0.7   3.0   0.6
  2.|-- 185.91.11.98             0.0%   100    0.3   1.2   0.2  32.8   4.3
  3.|-- 185.91.10.122            0.0%   100   48.4  41.7  36.3  61.6   6.5
        185.91.10.118          
  4.|-- 185.91.11.161            0.0%   100   49.6  56.2  48.9  75.2   5.1
        185.91.10.120          
  5.|-- 103.35.125.43            0.0%   100   48.4  56.8  48.4  86.4   7.1
        185.91.11.161       
```   
There's enough information here to deduce :
1. Current path is a 5 HOP path since dst (103.35.125.43 ) is the Primary Host at HOP - 5
But we do not know what the old path was, we can easily tweak the display function to print something like:
```
root@eqx-sjc-engine1-staging:~/atib_custom_mtr/mtr# ./mtr 103.35.125.43 -n --report-wide -c 100
Start: 2020-08-25T10:36:56+0000
HOST: eqx-sjc-engine1-staging Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 103.35.127.1             0.0%   100    2.9   4.3   0.7 137.2  14.5
  2.|-- 185.91.11.98             0.0%   100    0.8   0.7   0.2  14.9   2.1
  3.|-- 185.91.10.122            0.0%   100   48.6  42.5  36.3  55.8   6.5
        185.91.10.118          
  4.|-- 185.91.11.161            0.0%   100   50.3  55.5  48.9  70.9   5.1
        185.91.10.120          
  5.|-- 103.35.125.43            0.0%   100   48.4  55.5  48.4  70.3   6.0
        185.91.11.161          
  6.|-- 103.35.125.43            0.0%    57   53.3  58.9  53.3  59.1   0.8
```
There's enough information here to deduce :
1. Current path is a 5 HOP path since dst (103.35.125.43 ) is the Primary Host at HOP - 5
2. Old path was 6 HOP old since dst was a Primary Server at HOP - 6
3. Path Changed after 57 packets were already sent out

**Comment by rewolff** [comment]
OK. Good stuff! Nice work! 

On the todo list: Now in your case both before and after the route change you were getting 100% of  the probes back. But what if "old path had 10% packetloss" and "new path has 20% packetloss". 

What I'm thinking is that we track the "sent" with each HOST-at-hop-X. Just a simple heuristic: If this host was the last to respond, then the sent is accounted to him. 

So now 
```
5.|-- 103.35.125.43           10.0%   57   48.4  55.5  48.4  70.3   6.0
        185.91.11.161           20.%    43    48.4  55.5  48.4  70.3   6.0
```
Oh... The other statistics should then also be "attributed" to one or the other host... 
Do you think it is feasable to add this eventually?

**Comment by atibdialpad** [comment]
Yeah, that is a very good candidate for the todo list. 
So today, this (https://github.com/traviscross/mtr/blob/af2d385e9152bedf754e64eb3b34bd5b1caffbee/ui/net.c#L49) structure is used to store PER HOP stats, we will need to change this so that stats could be made PER HOP PER HOST! That does increase the size of the structure by roughly 8.(https://github.com/traviscross/mtr/blob/ace8ef7f5279ace5417de8b034d0a85e8f3ff313/ui/mtr.h#L64) times.
Not sure if that will have implications.

**Comment by rewolff** [comment]
IF that starts out as 1k, and becomes 8k per-hop, so 240k for 30 hops.... That's still acceptable in modern times. With a web browser window taking gigabytes that quarter-megabyte is acceptable. 

The goal is to write clear and easy code. So with it being very unlikely that you are going to see 8 different hosts at a certain hop, we can probably save 200k of that 240k by allocating those records dynamically. But that's going to complicate the code. The code should be as easy-to-read as possible. No distractions. 

A test-mtr is taking 9.3Mb of memory. You're going to increase that to 9.5 just for a simple extra feature. Don't worry about it. 

(And I'm probably over-estimating the extra memory use by a factor of ten already!)

**Comment by atibdialpad** [comment]
thanks for the analysis.. makes perfect sense. I shall add it to the TODO list when I raise the PR.

**Comment by rewolff** [comment]
I worked on a  project that had to be "real-time": We had a 25 ms deadline that we had to meet every time. When I got the project each of the 10 sub-modules was heavily optimized assembler code. But each module had a bug. While debugging I eventually rewrote every module in easy-to-read C. Now the full  calculation took around 50ms: Miss the deadline by quite some margin! Then I optimized PART of one of the modules and then we were suddenly able to meet the deadline: That one routine, unoptimized was taking 180% of the allowed time. Optimized about 60%. The other 9 modules used only 20% total. i.e. in that project only ONE routine needed optimization, all the others should be flexible and as easy-to-maintain as possible. 

So... in this case here. Readability and maintainability are MUCH more important than thinking about wasting a few bytes. Wasting gigabytes of unnecessary memory: not good, but a few hundred kb here and there is fine if it keeps the code nice and clean.
